### PR TITLE
feat: implement batching [VM-614]

### PR DESCRIPTION
**Pull request by justprosh**

### CODE REVIEW

File: crates/chain-listener/src/proof_tracker.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    pub async fn confirm_proof(&mut self, cu_id: CUID) -> U256 {
+        let proof_id = Uint::from(
+            *self
+                .proof_counter
+                .entry(cu_id)
+                .and_modify(|c| {
+                    *c = c.add(Uint::from(1));
+                })
+                .or_insert(Uint::from(1)),
```
Comment (monoid) [inline crates/chain-listener/src/proof_tracker.rs:96]:
```
let proof_count = self.proof_counter.entry(cu_id).or_insert(Uint::from(0));
*proof_count += 1;
let proof_id = Uint::from(*proof_count);
```
would be more explicit.

Comment (monoid) [reply crates/chain-listener/src/proof_tracker.rs:96]: The top-level `Uint::from`, is it really needed?

File: crates/chain-listener/src/listener.rs [hunk 2]
```diff
@@ -1075,72 +1041,130 @@ impl ChainListener {
...
-        let result_hash = ResultHash::from_slice(*self.difficulty.as_ref());
-
-        // proof_id is used only by CCP and is not sent to chain
-        let proof_id = CCProofId::new(self.global_nonce, self.difficulty, ProofIdx::zero());
-        let units = self.cc_compute_units.keys().cloned().collect::<Vec<_>>();
-        for unit in units {
-            let local_nonce = LocalNonce::random();
-            self.submit_proof(CCProof::new(proof_id, local_nonce, unit, result_hash))
-                .await?;
+    // /// Submit Mocked Proofs for all active compute units.
```
Comment (monoid) [inline crates/chain-listener/src/listener.rs:1044]: More commented code.

File: crates/chain-listener/src/listener.rs [hunk 3]
```diff
@@ -1075,72 +1041,130 @@ impl ChainListener {
...
+            let proof_batches = if self.is_epoch_ending() {
+                let last_known_proofs = batch_requests
+                    .into_iter()
+                    .map(|(cu_id, req)| (cu_id, req.last_seen_proof_idx))
+                    .collect();
+
+                tracing::debug!(target: "chain-listener", "Polling proofs after {:?}", last_known_proofs);
+                measured_request(
+                    &self.metrics,
+                    async { ccp_client.get_proofs_after(last_known_proofs, PROOF_POLL_LIMIT) }
```
Comment (monoid) [inline crates/chain-listener/src/listener.rs:1098]: The `async { X }.await` construction may be replaced by just `X`.  Here and elsewhere.

File: crates/chain-connector/src/connector.rs [hunk 4]
```diff
@@ -806,6 +816,14 @@ impl ChainConnector for HttpChainConnector {
     }
 }
 
+fn parse_str_field(value: Option<Value>, field: &str) -> eyre::Result<String> {
+    value
+        .ok_or(eyre!("Field {} not found in response", field))?
```
[the remark this answers is not in the document]
Comment (gurinderu) [reply crates/chain-connector/src/connector.rs:824]: it should be enum without eyre)

[the remark this answers is not in the document]
Comment (monoid) [reply crates/chain-connector/src/connector.rs:824]: Even better.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -11,6 +11,10 @@ on:
         description: "Cargo dependencies map"
         type: string
         default: "null"
+    secrets:
```
Comment (gurinderu) [inline .github/workflows/build.yml:14]: Why did you change that in this pr?

Comment (justprosh) [reply .github/workflows/build.yml:14]: it was done by @enjenjenje because of private repo deps

File: crates/chain-connector/src/connector.rs [hunk 6]
```diff
@@ -468,25 +473,28 @@ impl HttpChainConnector {
             gas_limit: Uint256::from_le_bytes(&gas_limit.to_le_bytes_vec()),
             to: to.parse()?,
             value: 0u32.into(),
-            data,
+            data: data.clone(),
```
Comment (gurinderu) [inline crates/chain-connector/src/connector.rs:479]: Data can be a pretty big, probably it is better to apply hex::decode before this line and do not use cloning ?

File: crates/chain-listener/src/proof_tracker.rs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            self.last_submitted_proof_ids.clear();
+            self.persist().await;
+            true
+        } else {
+            false
+        }
+    }
+
+    async fn persist(&self) {
+        let backoff = ExponentialBackoff {
```
Comment (gurinderu) [inline crates/chain-listener/src/proof_tracker.rs:134]: Why do we need a backoff for the persistence ?

File: crates/chain-data/src/block_header.rs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            .ok_or(eyre::eyre!(" timestamp field not found; got {json}"))?
+            .to_string();
+
+        let block_number = obj
+            .get("number")
+            .and_then(Value::as_str)
+            .ok_or(eyre::eyre!("number field not found; got {json}"))?
+            .to_string();
+
+        Ok(Self {
```
Comment (monoid) [inline crates/chain-data/src/block_header.rs:45]: The type may be deserialized from a `Value` with serde in more declarative manner.
https://docs.rs/serde_json/latest/serde_json/fn.from_value.html
